### PR TITLE
Sync validation

### PR DESCRIPTION
**Pull request by lazcool**

### CODE REVIEW

File: DFC.ServiceTaxonomy.GraphSync/GraphSyncers/GraphSyncHelper.cs [hunk 1]
```diff
@@ -61,38 +61,50 @@ public async Task<IEnumerable<string>> NodeLabels()
...
+            return await NodeLabels(graphSyncPartSettings, contentType);
+        }
+
+        private async Task<IEnumerable<string>> NodeLabels(GraphSyncPartSettings graphSyncPartSettings, string contentType)
+        {
+            string nodeLabel = graphSyncPartSettings.NodeNameTransform switch
+            {
+                "$\"ncs__{ContentType}\"" => $"ncs__{contentType}",
+                "$\"esco__{ContentType}\"" => $"esco__{contentType!}",
+                _ => await TransformOrDefault(graphSyncPartSettings.NodeNameTransform, contentType, contentType)
```
Comment (tomtollafield) [inline DFC.ServiceTaxonomy.GraphSync/GraphSyncers/GraphSyncHelper.cs]: Out of interest more than anything else, what does this do?

File: DFC.ServiceTaxonomy.GraphSync/GraphSyncers/Parts/BagPartGraphSyncer.cs [hunk 2]
```diff
@@ -68,11 +69,10 @@ public BagPartGraphSyncer(IServiceProvider serviceProvider)
                 delayedReplaceRelationshipsCommand.SourceIdPropertyName = mergeNodeCommand.IdPropertyName;
                 delayedReplaceRelationshipsCommand.SourceIdPropertyValue = (string?)mergeNodeCommand.Properties[delayedReplaceRelationshipsCommand.SourceIdPropertyName];
 
-                //todo: move this code into graphSyncHelper
-                //todo: what if want different relationships for same contenttype in different bags!
-                string? relationshipType = graphSyncHelper.GraphSyncPartSettings.BagPartContentItemRelationshipType;
-                if (string.IsNullOrEmpty(relationshipType))
-                    relationshipType = await graphSyncHelper.RelationshipType(contentType);
+                var bagContentItemGraphSyncHelper = _serviceProvider.GetRequiredService<IGraphSyncHelper>();
```
Comment (tomtollafield) [inline DFC.ServiceTaxonomy.GraphSync/GraphSyncers/Parts/BagPartGraphSyncer.cs]: Can you not use normal DI via constructor for this?

File: DFC.ServiceTaxonomy.GraphSync/GraphSyncers/Parts/BagPartGraphSyncer.cs [hunk 3]
```diff
@@ -86,32 +86,59 @@ public BagPartGraphSyncer(IServiceProvider serviceProvider)
...
+            foreach (ContentItem bagPartContentItem in contentItems)
             {
-                //todo: foreach will never iterate
-                if (await graphSyncValidator.CheckIfContentItemSynced(bagPartContentItem))
-                {
-                    return true;
-                }
-                else
-                {
+                var graphSyncValidator = _serviceProvider.GetRequiredService<IGraphSyncValidator>();
```
Comment (tomtollafield) [inline DFC.ServiceTaxonomy.GraphSync/GraphSyncers/Parts/BagPartGraphSyncer.cs]: Same here, seeing it twice though I'm sure there's a reason

File: DFC.ServiceTaxonomy.GraphSync/Activities/AuditSyncIssuesTask.cs [hunk 4]
```diff
@@ -1,20 +1,20 @@
 using System;
 using System.Collections.Generic;
-using System.Linq;
+// using System.Linq;
```
[the remark this answers is not in the document]
Comment (lazcool) [reply DFC.ServiceTaxonomy.GraphSync/Activities/AuditSyncIssuesTask.cs]: They're the `using`s required for the commented out code in the file. The first thing I'll do in the next pr is uncomment everything again 😄.

File: DFC.ServiceTaxonomy.GraphSync/Activities/AuditSyncIssuesTask.cs [hunk 5]
```diff
@@ -57,60 +57,73 @@ public override IEnumerable<Outcome> GetPossibleOutcomes(WorkflowExecutionContex
...
+            // var log = (await _session.Query<AuditSyncLog>().FirstOrDefaultAsync()) ??
+            //           new AuditSyncLog {LastSynced = DateTime.MinValue};
+            //
+            // //todo: do we want to load 10s thousands at once??
+            // var contentItems = await _session
+            //     //do we only care about the latest published items?
+            //     .Query<ContentItem, ContentItemIndex>(x =>
+            //         x.ContentType.IsIn(contentTypes.Keys) && x.Latest && x.Published && (x.CreatedUtc >= log.LastSynced ||  x.ModifiedUtc >= log.LastSynced))
+            //     .ListAsync();
+            //
```
[the remark this answers is not in the document]
Comment (lazcool) [reply DFC.ServiceTaxonomy.GraphSync/Activities/AuditSyncIssuesTask.cs]: Yeah, the commenting out effectively makes the sync validation triggered by the workflow a no-op.

File: DFC.ServiceTaxonomy.GraphSync/CSharpScripting/ContentHelper.cs [hunk 6]
```diff
@@ -33,8 +33,13 @@ public async Task<string> GetContentItemIdByDisplayText(string contentType, stri
 
             ContentItem contentItem = await query.FirstOrDefaultAsync();
 
-            if (contentItem == null) //todo: what exception?
-                throw new InvalidDataException($"Unable to get content item of type '{contentType}' with DisplayText '{displayText}'.");
+            if (contentItem == null)
+            {
+                #pragma warning disable S112
+                //todo: best exception?
+                throw new Exception($"Unable to get content item of type '{contentType}' with DisplayText '{displayText}'.");
```
[the remark this answers is not in the document]
Comment (lazcool) [reply DFC.ServiceTaxonomy.GraphSync/CSharpScripting/ContentHelper.cs]: Introduced a new `ContentHelperExtension` and throw that instead. (The `ContentHelper` class may be used outside of c# scripting.)

File: DFC.ServiceTaxonomy.GraphSync/GraphSyncers/GraphSyncHelper.cs (hunk 1 above)
Comment (lazcool) [reply DFC.ServiceTaxonomy.GraphSync/GraphSyncers/GraphSyncHelper.cs]: The switch expression checks for the actual `NodeNameTransform` config values we use for NCS and ESCO nodes, and if it finds them, it creates the node label without invoking Roslyn. This is to speed up the import process.
If the transform is not one of the current values we use in the startup recipe, it executes the transform code snippet to generate the node label.

File: DFC.ServiceTaxonomy.GraphSync/GraphSyncers/Parts/BagPartGraphSyncer.cs (hunk 2 above)
Comment (lazcool) [reply DFC.ServiceTaxonomy.GraphSync/GraphSyncers/Parts/BagPartGraphSyncer.cs]: We need an `IGraphSyncHelper` for each content item in the bag. We only know how many items are in the bag during the sync process, not at construction time.

**Pull request merged**